### ISSUE TITLE
FabGL Edition BBC BASIC for ESP32

### ISSUE THREAD

**Issue by OzHawk** [rtrussell/BBCSDL#12]
This issue captures the possibility of creating a FabGL edition of the 32bit BBC BASIC for the ESP32.  

Given that this project is called BBCSDL, it may not be the best place for it, but looking at the commonality of files, it is best kept with the rest of the project.

This suggestion is to implement a version of BBC BASIC that runs on a FabGL ESP32 with PS/2 keyboard (and mouse?), VGA output, audio out, and SD card.  Additional features could include wifi networking.

Effort required to replace SDL calls with FabGL calls, including screen management.
There is also significant effort required to implement the ASM for ESP32.

Reference: [https://github.com/fdivitto/FabGL]

**Comment by ejolson2005** [comment]
There is also a Pico version with VGA that is not in the master repository.  It currently works well enough to do some simple graphics and will hopefully become feature complete soon.  At any rate, you may want to look at the code in Bill's repository (Memotech-Bill under forks) for a cross between the SDL and console versions to start with for the FabGL edition.

**Comment by OzHawk** [comment]
It looks like this activity is being tackled from a couple of directions.
[https://github.com/fdivitto/FabGL/discussions/154](url)

**Comment by OzHawk** [comment]
@rtrussell , would you like to see an ESP32 version merged in with your SDL version, or would you rather it was a seperate project?
Is it possible to extract out the common parts as a seperate project and the SDL, Pico and ESP32 version include that?

**Comment by rtrussell** [comment]
> would you like to see an ESP32 version merged in with your SDL version, or would you rather it was a seperate project?

I don't think it would be straightforward to merge a non-SDL GUI version (such as the VGA version for the Pico that Bill is working on) with the BBCSDL repository, since it's significantly different in architecture from both the existing SDL2 and Console Mode versions.

I know that Bill has raised a Pull Request to do exactly that, but I don't know how he has tried to resolve those differences; it would be interesting to see a diagram showing the architecture.  I won't be merging that Pull Request without a lot of careful thought, because I'm pretty sure that as it stands it would impact other editions.

I would suggest you talk to Bill to establish whether his Pico/VGA build can be adapted for the ESP32 and if so how that might impact on how he has structured it.

**Comment by OzHawk** [comment]
Thanks Richard.  The Pico and ESP32 versions will have more in common with each other than with the SDL versions.  That is why I asked the question.
An ESP32 console version would have a lot in common with the Pico version.
I am currently assessing the library requirements for an ESP32 version. I was planning on using the FabGL library for the VGA and PS/2 portions of the build, but they seem to be dependant on an Arduino IDE.  I need to work out how to convert that to a Makefile.  Same with the ESP-IDF.
Maybe I’m best to work in my own fork initially and merge changes in after I’ve gotten my head around the requirements and architecture more.

**Comment by rtrussell** [comment]
> An ESP32 console version would have a lot in common with the Pico version.

It would be very nice if only one new file - substituting for **bbpico.c** - needs to be written for the ESP32.  I would expect, and hope, that the various places where there's an **#ifdef PICO** in the other source files are equally applicable to any microcontroller (so you would have to specify **-DPICO** in your makefile).

From my perspective it's the console edition that is really useful, because it allows you to configure the board as a standalone controller (Eric always gives the example of a washing machine programmed in BBC BASIC, but that's probably not very realistic!).  A GUI/VGA version feels more of a gimmick to me: that it's possible at all is amazing, but probably of limited practical use.

**Comment by OzHawk** [comment]
This sounds like a good place for me to start.  My early learning will be about developing the Makefile to build an ESP32 version and what libraries I need to include.

I'm very interested in a VGA version of BBC BASIC for ESP32 as I'm looking for a BASIC OS that I can use for a FabGL computer.  I want to document how to build a FabGL compatible computer from breadboard, all the way up to a custom PCB.  The idea is that kids could make their own Modern Retro Commputer and learn to program it much like we did in the old days.  Much like the Colour Maximite computer, only cheaper.  Sure, less capable, but can be made using parts that people are currently throwing away and a microcontroller dev board less than $10.

A Pico version could be a viable alternative, as they are also very inexpensive when compared to the Waveshare boards that the Colour Maximite use.

I looked at using Paul Scott Robinson's Eris BASIC, however it was designed around his own virtual machine which made it very difficult to expend further.

I would like to publish something like this in a local Australian electronics magazine called "Diyode".  

Anyway, that's my plan.  I'm also working towards developing a 3D printable case that mounts a keyboard and the mirocontroller much like the Pi 400.  Why not just buy a Pi 400?  That's a great question, as it could be made to do what I'm after, but it's more about the ability to make your own.  I am currently making one on prototype borad from mostly recovered parts in an old Digital TV tuner case to drive an old 15" screen.  I want others to experience the enjoyment of creating it themselves.

Anyway, that is my motives for a full VGA version.
